### PR TITLE
Multivariate Timeseries Input

### LINKED ISSUE #87: dynamic scalability of Cyclegan primitive based on `window_size`
In the current implementation of cyclegan, changing the `window_size` might not successfully work since units the layers are independent. 

For example the encoder layer is defined as:
```
"layers_encoder": {
                "type": "list",
                "default": [
                    {
                        "class": "keras.layers.Bidirectional",
                        "parameters": {
                            "layer": {
                                "class": "keras.layers.LSTM",
                                "parameters": {
                                    "units": 100,
                                    "return_sequences": true
                                }
                            }
                        }
                    },
                    {
                        "class": "keras.layers.Flatten",
                        "parameters": {}
                    },
                    {
                        "class": "keras.layers.Dense",
                        "parameters": {
                            "units": 20
                        }
                    },
                    {
                        "class": "keras.layers.Reshape",
                        "parameters": {
                            "target_shape": "encoder_reshape_shape"
                        }
                    }
                ]
            }
```

In this case, the units of the LSTM layers should equal to the input size (`window_size`).

### LINKED ISSUE #248: Supporting multivariate input
In addition to univariate signals, we would like to include multivariate signals as possible inputs to the pipeline. The general format of the data will follow the following structure

|    timestamp   |     dim 0      |     dim 1       |     dim 2      |
| :------------: | :----------: | :----------: | :----------: |
| 1222819200  | -0.366358 |         0          |         0          |
| 1222840800 | -0.394107  |         1          |         0          |
| 1222862400 |  0.403624  |         1          |         1          |
| 1222884000 | -0.362759 |         0          |         0          |
| 1222905600 | -0.370746  |         1          |         0          |

The usage of the Orion class will stay the same, we need to only alter the variable `input_shape` to allow multiple channel ingestion. For now, we assume that we are only predicting / reconstructing a single channel (MISO).
```python3
from orion import Orion 
 
 hyperparameters = { 
     'orion.primitives.tadgan.TadGAN#1': { 
         'input_shape': [100, 3], # where 100 is the window size and 3 is the number of channels 
         'verbose': True 
     } 
} 

orion = Orion( 
     pipeline='tadgan', 
     hyperparameters=hyperparameters 
) 
```

### PR DESCRIPTION
**Pull request by sarahmish**
Resolve #87.
Resolve #248.

In 'setuptools>58.0.0' [release](https://setuptools.pypa.io/en/latest/history.html#v58-0-0), `lib2to3` is no longer supported making `mongoengine<0.20` fail. Upgrading `mongoengine>=0.20`.

### CODE REVIEW

File: docs/user_guides/primitives_pipelines/primitives/TadGAN.rst [hunk 1]
```diff
@@ -15,21 +15,18 @@ argument                    type                description
 **parameters**
 ------------------------------------------------------------------------------------------------------------------------------------------------
  ``X``                      ``numpy.ndarray``   n-dimensional array containing the input sequences for the model 
-
+ ``y``                      ``numpy.ndarray``   n-dimensional array containing the target sequences we want to reconstruct
```
Comment (dyuliu) [inline docs/user_guides/primitives_pipelines/primitives/TadGAN.rst:18]: Is ``y`` always the same as ``X``? If yes, we can add some quick explanations here.

File: docs/user_guides/primitives_pipelines/primitives/TadGAN.rst [hunk 2]
```diff
@@ -51,11 +48,11 @@ argument                    type                description
 
     X = np.array([1] * 100).reshape(1, -1, 1)
     primitive = load_primitive('orion.primitives.tadgan.TadGAN', 
-        arguments={"X": X, "epochs": 5, "batch_size": 1,
+        arguments={"X": X, "y":X, "epochs": 5, "batch_size": 1,
                    "iterations_critic": 1})
 
     primitive.fit()
-    y, critic = primitive.produce(X=X)
+    y, critic = primitive.produce(X=X, y=X)
```
Comment (dyuliu) [inline docs/user_guides/primitives_pipelines/primitives/TadGAN.rst:55]: Is there any necessary reason to expose y for the reconstruction-based method?

File: notebooks/Using Orion on multivariate input.ipynb [hunk 3]
```diff
@@ -0,0 +1,528 @@
+{
```
Comment (dyuliu) [inline notebooks/Using Orion on multivariate input.ipynb:1]: Did not fully check this file. But I know there is a way to automatically test code snippets in notebooks (refer to sdv library). Not sure whether it is worthwhile to have a similar one here.

File: orion/pipelines/verified/dense_autoencoder/dense_autoencoder.json [hunk 4]
```diff
@@ -4,6 +4,7 @@
         "sklearn.impute.SimpleImputer",
         "sklearn.preprocessing.MinMaxScaler",
         "mlprimitives.custom.timeseries_preprocessing.rolling_window_sequences",
+        "orion.primitives.timeseries_preprocessing.slice_dim",
```
Comment (dyuliu) [inline orion/pipelines/verified/dense_autoencoder/dense_autoencoder.json:7]: What's the function of ``slice_dim``? I can tell it from the name.

File: orion/primitives/timeseries_preprocessing.py [hunk 5]
```diff
@@ -2,6 +2,38 @@
 import pandas as pd
 
 
+def slice_dim(X, target_index, axis=0):
```
Comment (dyuliu) [inline orion/primitives/timeseries_preprocessing.py:5]: slice_array?  slice_array_by_dims?

File: docs/user_guides/primitives_pipelines/primitives/TadGAN.rst (hunk 1 above)
Comment (sarahmish) [reply docs/user_guides/primitives_pipelines/primitives/TadGAN.rst:18]: `y` is a selected channel of `X`. So it can be `X` or a few channels from `X`. I can add that to the documentation.

File: docs/user_guides/primitives_pipelines/primitives/TadGAN.rst (hunk 2 above)
Comment (sarahmish) [reply docs/user_guides/primitives_pipelines/primitives/TadGAN.rst:55]: Yep, sometimes we do not reconstruct all of `X` but only a selected subset of channels. In this example, we are using a univariate signal, that's why they are the same. 

I prefer to expose them rather than let the user specify the dim, because it makes it easier to retrieve the signal for debugging or visualization purposes.

File: notebooks/Using Orion on multivariate input.ipynb (hunk 3 above)
Comment (sarahmish) [reply notebooks/Using Orion on multivariate input.ipynb:1]: Good suggestion, I will add it in another PR to keep them separate. 
Follow #251.
